### PR TITLE
[FIX] rename_xmlids: don't merge if hierarchy exists

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
When we already have recursive hierarchy, doing a merge of a parent into one of their children let the awkward situation where the child points to itself, so we avoid it checking this condition.

Same as done in https://github.com/OCA/OpenUpgrade/pull/2730.

### CODE REVIEW

**Review by legalsylvain — APPROVED** [review]
Lgtm. Thanks.

**Comment by pedrobaeza** [comment]
I personally see this as a very special case for being handled in such generic rename method, but I let @hbrunn and @StefanRijnhart to give their opinion.

**Review by hbrunn — APPROVED** [review]

**Comment by StefanRijnhart** [comment]
Wouldn't it be better to rewire the hierarchy? Is there a concrete case?

**Comment by pedrobaeza** [comment]
This is coming from https://github.com/OCA/OpenUpgrade/pull/2730

**Comment by StefanRijnhart** [comment]
OK, so this is the more generalized case. But the merging is executed by `merge_records`, so shouldn't the intervention be in there, and again, shouldn't the proper solution be to do the merge but prevent the circular reference?

**Comment by kos94ok-3D** [comment]
I think better use?
image

**Comment by pedrobaeza** [comment]
Yes, the field name shouldn't be hard-coded.

**Comment by StefanRijnhart** [comment]
Maybe you will want to check by FK constraint instead of `_parent_name`, so that you catch Many2many references as well, such as res.group's implied_ids.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Pending comments

**Comment by MiquelRForgeFlow** [comment]
@StefanRijnhart your comments are attended. The PR it's ready :)

File: openupgradelib/openupgrade_merge_records.py [hunk 1]
```diff
@@ -682,6 +682,53 @@ def _delete_records_orm(env, model_name, record_ids, target_record_id):
...
+        if new_parent_row and new_parent_row[0] in record_ids:
+            # When we already have recursive hierarchy, doing a
+            # merge of a parent into one of their children let the
+            # awkward situation where the child points to itself,
+            # so we avoid it checking this condition
+            logger.info(
+                "Couldn't merge %s record(s) of model %s to record_id %s"
+                " to avoid recursion with field %s of table %s",
+                len(record_ids), model_name, target_record_id, origin, table)
+            return True
```
Comment (StefanRijnhart) [inline openupgradelib/openupgrade_merge_records.py:728]: Thanks for the refactoring. Now I'm just concerned with the semantics: I would expect a method called "check_no_recurrence" to return True if there is indeed "no recurrence" but I believe that is the other way around here, right?

Comment (MiquelRForgeFlow) [reply openupgradelib/openupgrade_merge_records.py:728]: hahaha Ok, I renamed it to "_check_recurrence"

**Review by StefanRijnhart — APPROVED** [review]
Thanks!

**Comment by pedrobaeza** [comment]
I see this for detecting problems, but I think we should provide a way to auto-fix these ones, and maybe doing it through `adjust_values` (auto remove m2m references to itself for example) , or should we run it a time to see real problems?

**Comment by MiquelRForgeFlow** [comment]
> or should we run it for some time to see real problems?

Yes, do that in the future, not now. At least until we know for sure we can do that auto remove without worries.

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
I shouldn't merge this with such bad commit message, but there's no solution now. I will put at least a good description on the release message.

**Pull request merged**